packages/@lwc/template-compiler/src/parser/parser.ts

```diff
@@ -45,6 +48,32 @@ interface ParentWrapper {
     current: ParentNode;
 }
 
+interface IfContext {
+    currentNode: IfBlock | ElseifBlock | ElseBlock;
+
+    // Within a specific if-context, each set of seen slot names must be tracked separately
+    // because duplicate names in separate branches of the same if-block are allowed (the branching
+    // logic provides a compile-time guarantee that the slots will not be rendered multiple times).
+    // seenSlots keeps track of each set holding the slots seen in each branch of the if-block.
+    seenSlots: Set<string>[];
+}
+
+/**
+ * A SiblingScope object keeps track of the context needed to parse a series of if-elseif-else nodes.
+ *
+ * @param {IfContext} ifContext Context for the if-elseif-else chain currently being parsed at this level. This
+ * IfContext keeps track of the most recently parsed node in the chain and the set of slot names we've seen in all
+ * previous siblings in the chain.
+ * @param {IfContext} ancestorIfContext Reference to the nearest ancestor IfContext. The existence of an ancestor
+ * IfContext means that we are currently parsing nodes nested within an if-elseif-else chain. Context from that ancestor
+ * is needed to track which slot names have already been seen in and only in the current scope. This reference is also needed
+ * so we know where to merge all visited slot names from the current IfContext.
```

Comment:
The current JSDOC is invalid, the `@param` is revered to function arguments. I would suggest moving the comments inline above each field.